### DISCUSSION TITLE
question about chart timestamp format

### DISCUSSION THREAD

**Discussion by timmpo** [frangoteam/FUXA#1458] [General]
i have tried to send coustom data to line chart, but have trouble to get time series axis to work,
when paramLines return x axis in unix-time (seconds) it plots the data but need some conversion to human readable format, i have tried to send timestamps in string format etc (2024/10/10-17:50) but it dont works.
what is the correct time format to send ? 

working example of paramLines returning values, but need some human readable conversion:

```typescript
 {
    id: 't_b070199c-92384f8b',
    name: 'chartDataPoint',
    device: 'FUXA Server',
    color: '#4484ef',
    label: 'chartDataPoint',
    yaxis: 1,
    x: [
      1730022583, 1730022593,
      1730022603, 1730022613,
      1730022623, 1730022633,
      1730022643, 1730022653,
      1730022663, 1730022673,
      1730022683, 1730022693,
      1730022703, 1730022713,
      1730022723, 1730022733,
      1730022743, 1730022753
    ],
    y: [
      93, 93, 78, 57, 37, 52, 49,
      16, 11, 35, 98, 39, 91, 55,
      53, 60, 66, 51
    ]
  }
]
```

**Comment by unocelli** [comment]
Hi, you need timestamp in milliseconds:

const dateStr = "2024-10-31T15:30:00";
const dateObj = new Date(dateStr); 
const timestamp = dateObj.getTime(); // milliseconds

**Reply by timmpo** [reply]
I have tested milliseconds, and the chart unfortunately not convert it. 

i send this to the chart: 
```
[
  {
    id: 't_b070199c-92384f8b',
    name: 'chartDataPoint',
    device: 'FUXA Server',
    color: '#4484ef',
    label: 'chartDataPoint',
    yaxis: 1,
    x: [
      1730388513000, 1730388518000, 1730388523000,
      1730388528000, 1730388533000, 1730388538000,
      1730388543000, 1730388548000, 1730388553000,
      1730388558000, 1730388563000, 1730388568000,
      1730388573000, 1730388578000, 1730388583000,
      1730388588000, 1730388593000, 1730388598000,
      1730388603000, 1730388608000, 1730388613000,
      1730388618000, 1730388623000, 1730388628000,
      1730388633000, 1730388638000, 1730388643000,
      1730388648000, 1730388653000, 1730388658000,
      1730388663000, 1730388668000, 1730388673000,
      1730388678000, 1730388683000, 1730388688000
    ],
    y: [
      23, 53, 29, 18, 25, 45, 49, 100, 66,
      84, 20, 58, 39, 78, 13, 45,  52,  0,
      71, 44, 53, 31, 18, 18, 49,  51, 84,
      36, 34, 87, 76, 78, 63,  6,  22, 27
    ]
  }
]
``` 

and the result: 
![datapoint](https://github.com/user-attachments/assets/64dd297d-8643-40a0-9950-0f9d42029dd6)

**Comment by timmpo** [comment]
im maybe out in the woods here, but if no options for date format is send to uplot, maybe uplot triggers the rawValue in checkDateFormat() in ngx-uplot.component.ts. if that the case it will be a good feature to have the option to chose in Chart settings menu under date format. a option to convert timestamps to dates/times or treat x values as plain numbers for custom charts.

**Reply by unocelli** [reply]
Hi @timmpo, I think you are right, the feature is missing!

**Comment by timmpo** [comment]
in file ngx-uplot.component.ts where init(options?: ChartOptions, rawData?: boolean) we needed options from chart setting to enable/disable rawData or view as plain number.

in chart-uplot.component.ts where public setCustomValues(lines) we cant convert millis to seconds if we want to use
coustom x values instead of timestamps so some form of detection needed there.


i have made a simple hack for those who really need to send coustom chart with timestamps for now.
The timestamps must be send as seconds and not millis for coustom chart as for now,
it checks the x axis label name if it has changed from the default value "Time" as we dont want that label anyway if we use other then timestamps for x axis data.
in ngx-uplot.components.ts line 185 i added

```
		if (options.axes && options.axes[0].label && options.axes[0].label === "Time") {
			rawData = false;
			this.rawData = false;
		}
``` 
and then rebuild
